Frameworks/Foundation/NSKVOSupport.mm

```diff
@@ -516,169 +515,301 @@ - (void)removeObserver:(id)observer forKeyPath:(NSString*)keyPath {
+            _NSKVOKeypathObserver* keypathObserver = keyObserver.keypathObserver;
+            NSKeyValueObservingOptions options = keypathObserver.options;
+            id rootObject = keypathObserver.object;
+            id observer = keypathObserver.observer;
+            NSString* keypath = keypathObserver.keypath;
+            NSMutableDictionary* change = keypathObserver.pendingChange;
+            void* context = keypathObserver.context;
 
-        if (![keypathObserver popDidChange]) {
-            // Skip any keypaths that are in the process of changing.
-            continue;
-        }
+            if ((options & NSKeyValueObservingOptionNew) && [change[NSKeyValueChangeKindKey] integerValue] != NSKeyValueChangeRemoval) {
+                // For to-many mutations, we can't get the new values at indexes that have been deleted.
+                id newValue = keypathValueCache[keypath];
+                if (!newValue) {
+                    newValue = [rootObject valueForKeyPath:keypath] ?: [NSNull null];
+                    keypathValueCache[keypath] = newValue;
+                }
 
-        NSKeyValueObservingOptions options = keypathObserver.options;
-        id rootObject = keypathObserver.object;
-        id observer = keypathObserver.observer;
-        NSString* keypath = keypathObserver.keypath;
-        NSMutableDictionary* change = keypathObserver.pendingChange;
-        void* context = keypathObserver.context;
-
-        if ((options & NSKeyValueObservingOptionNew) && [change[NSKeyValueChangeKindKey] integerValue] != NSKeyValueChangeRemoval) {
-            // For to-many mutations, we can't get the new values at indexes that have been deleted.
-            id newValue = [keypathValueCache objectForKey:keypath];
-            if (!newValue) {
-                newValue = _valueForPendingChange(notifyingObject, key, rootObject, keypath, keyObserver, change);
-                [keypathValueCache setObject:newValue forKey:keypath];
+                change[NSKeyValueChangeNewKey] = newValue;
             }
-            [change setObject:newValue forKey:NSKeyValueChangeNewKey];
-
-            // VSO 5051216: Implement set mutation notifications.
-        }
 
-        [observer observeValueForKeyPath:keypath ofObject:rootObject change:change context:context];
+            [observer observeValueForKeyPath:keypath ofObject:rootObject change:change context:context];
 
-        keypathObserver.pendingChange = nil;
+            keypathObserver.pendingChange = nil;
+        });
     }
 }
 
 /**
 @Status Interoperable
 */
-- (void)didChangeValueForKey:(NSString*)key {
-    if (![self observationInfo]) {
-        return;
+- (void)willChange:(NSKeyValueChange)changeKind valuesAtIndexes:(NSIndexSet*)indexes forKey:(NSString*)key {
+    if ([self observationInfo]) {
+        _dispatchWillChange(self, key, [&changeKind, indexes, key, notifyingObject = self ](_NSKVOKeyObserver * keyObserver) {
```

Comment:
This was to be able to modify it, which is apparently not the correct thing so will fix